### PR TITLE
Overhaul ParseConcentration

### PR DESCRIPTION
**Pull request by haydnKing**
I noticed that simulation times were really large for workflows with lots of DOE. CPU profiling indicated that we were spending a great deal of time (circa 50% of total in some cases) in `wunit.ParseConcentration`. 

This change simplifies the implementation of the function and more clearly defines its behaviour, extending the test cases to cover units like "% w/v" in the process.

For the bundle used during testing (the one attached to [ANTHA-1947](https://synthace.atlassian.net/browse/ANTHA-1947) ) simulation time is reduced from
```
real	2m4.954s
user	3m10.324s
sys	0m5.274s
```
to 
```
real	0m29.050s
user	0m43.102s
sys	0m3.842s
```

I'm running antha-tester right now, will post results

### CODE REVIEW

File: antha/anthalib/wunit/parse.go [hunk 1]
```diff
@@ -33,11 +34,53 @@ func extractFloat(s string) (float64, string) {
 	var longest int
 	var ret float64
 	for i := range s {
-		if f, err := strconv.ParseFloat(s[:i], 64); err == nil {
+		if f, err := strconv.ParseFloat(s[:i+1], 64); err == nil {
```
Comment (msackman-synthace) [inline antha/anthalib/wunit/parse.go]: lol, good point!

File: antha/anthalib/wunit/parse.go [hunk 2]
```diff
@@ -33,11 +34,53 @@ func extractFloat(s string) (float64, string) {
...
 	}
 
 	return ret, s[longest:]
 }
+
+// extractLastFloat extract the longest valid float from the right hand side of
+// the string and return the float and the remaining string.
+// If no float is found, returns zero and the entire string.
+func extractLastFloat(s string) (float64, string) {
+	longest := len(s)
```
Comment (msackman-synthace) [inline antha/anthalib/wunit/parse.go]: this should not be called `longest`. Maybe `startFrom` ?

**Comment by haydnKing** [comment]
I might have jumped the gun, looks like there are 3 test bundles failing... I'll have a look and post back

File: antha/anthalib/wunit/unitfromstring.go [hunk 3]
```diff
@@ -26,182 +26,71 @@ import (
...
-					//break
-				}
-			}
+	// unit location indicated by parentheses
+	if l, r := strings.Index(s, "("), strings.LastIndex(s, ")"); l < r {
+		inParentheses := strings.TrimSpace(s[l+1 : r])
+		value, unit := extractFloat(inParentheses)
+		if trimUnit := strings.TrimSpace(unit); !reg.ValidUnitForType("Concentration", trimUnit) {
+			return false, NewConcentration(0, "g/l"), s
+		} else if componentName := strings.TrimSpace(s[:l] + s[r+1:]); componentName != "" {
```
Comment (msackman-synthace) [inline antha/anthalib/wunit/unitfromstring.go]: if `)` is the last char/rune, then s[r+1:] will panic.

File: antha/anthalib/wunit/unitfromstring.go [hunk 4]
```diff
@@ -26,182 +26,71 @@ import (
...
 	}
 
-	componentNameOnly = strings.Join(notConcFields, " ")
+	approvedUnits := reg.ListValidUnitsForType("Concentration")
 
-	// if no match, return original component name
-	if unitmatchlength == 0 {
-		return false, conc, componentname
+	// value and unit at left
+	if value, remainder := SplitValueAndUnit(s); len(remainder) < len(s) { // value must be given for unit at left
```
Comment (msackman-synthace) [inline antha/anthalib/wunit/unitfromstring.go]: can you give examples in the comments of what will match here please?

File: antha/anthalib/wunit/unitfromstring.go [hunk 5]
```diff
@@ -26,182 +26,71 @@ import (
...
-
-	value, err := strconv.ParseFloat(concfields[0], 64)
-	if err != nil {
-		concfields[0] = strings.Trim(concfields[0], "()")
-		value, err = strconv.ParseFloat(concfields[0], 64)
-		if err != nil {
-			if concfields[0] == "" {
-				value = 0.0
+	// unit at right
+	if sym, remainder := extractLastSymbol(s, approvedUnits); sym != "" && remainder != "" {
```
Comment (msackman-synthace) [inline antha/anthalib/wunit/unitfromstring.go]: and also here please?

Reason being I'm trying to understand why in the former you grab the value first, then the symbol, whereas here you do it the other way around.

File: antha/anthalib/wunit/unitfromstring.go (hunk 3 above)
Comment (haydnKing) [reply antha/anthalib/wunit/unitfromstring.go]: I think we're safe because of the slice operator, https://play.golang.org/p/B_ln7-xmmXz

Comment (msackman-synthace) [reply antha/anthalib/wunit/unitfromstring.go]: head-off-by-one-error

File: antha/anthalib/wunit/parse.go (hunk 2 above)
Comment (m-i-sadowski) [reply antha/anthalib/wunit/parse.go]: either endAt or max, surely?

Comment (msackman-synthace) [reply antha/anthalib/wunit/parse.go]: sod it - `middleish` it is then.

or `startIfViewedFromTheNumberButEndIfViewedFromTheRemainder`. Hmm, not sure I miss Java...

Comment (msackman-synthace) [reply antha/anthalib/wunit/parse.go]: However, I reckon given the func is called `extractLastFloat`, it should be from the perspective of the float. So `startAt` makes sense. :P

**Comment by haydnKing** [comment]
Antha-tester is now running happily, so we're ready for another go

As @m-i-sadowski pointed out, wunit isn't the ideal home for this code, but since it's already there it seems out of scope to make larger changes

**Comment by msackman-synthace** [comment]
Full runs of antha-tester. Both passed.
Before:
```
real    11m28.769s
user    17m18.738s
sys     0m28.889s
```

After
```
real    7m56.191s
user    12m25.077s
sys     0m26.645s
```

**Comment by msackman-synthace** [comment]
Test coverage is excellent of the parse and unitfromstring files. Code looks good.

File: antha/anthalib/wunit/unitfromstring.go [hunk 6]
```diff
@@ -26,182 +26,70 @@ import (
...
-						notConcFields = append(notConcFields, fields[:i]...)
-					}
-					if len(fields) > i+1 {
-						notConcFields = append(notConcFields, fields[i+1:]...)
-					}
-					//break
-				}
-			}
+	// unit location indicated by parentheses - e.g. "Glucose (6M)"
+	if l, r := strings.Index(s, "("), strings.LastIndex(s, ")"); l < r {
```
[the remark this answers is not in the document]
Comment (msackman-synthace) [reply antha/anthalib/wunit/unitfromstring.go]: ah ha, even nicer: `0 <= l && l < r`

[the remark this answers is not in the document]
Comment (haydnKing) [reply antha/anthalib/wunit/unitfromstring.go]: oh good catch

File: antha/anthalib/wunit/unitfromstring_test.go [hunk 7]
```diff
@@ -136,6 +136,18 @@ func TestParseConcentration(t *testing.T) {
 			Conc:              NewConcentration(1.0, "X"),
 			ComponentNameOnly: "solutionX",
 		},
+		{
+			ComponentName:     "A happy solution :-) 5 g/l",
```
Comment (msackman-synthace) [inline antha/anthalib/wunit/unitfromstring_test.go]: lovely!

Comment (haydnKing) [reply antha/anthalib/wunit/unitfromstring_test.go]: :-)

**Pull request merged**